In [ ]:
import pandas as pd
time_metrics = pd.read_csv("/content/metrics_app_timeaggregated.csv")
print(time_metrics)

               id frequency                     start_time  \
0        20043506    hourly  2025-04-01 13:00:00.000 +0100   
1        20043507    hourly  2025-04-01 13:00:00.000 +0100   
2        20043508    hourly  2025-04-01 13:00:00.000 +0100   
3        20043509    hourly  2025-04-01 13:00:00.000 +0100   
4        20043510    hourly  2025-04-01 13:00:00.000 +0100   
...           ...       ...                            ...   
2411451  57501025    hourly  2025-07-09 14:00:00.000 +0100   
2411452  57501026    hourly  2025-07-09 14:00:00.000 +0100   
2411453  57501027    hourly  2025-07-09 14:00:00.000 +0100   
2411454  57501028    hourly  2025-07-09 14:00:00.000 +0100   
2411455  57501029    hourly  2025-07-09 14:00:00.000 +0100   

                              end_time metric_name aggregation  value  \
0        2025-04-01 13:59:59.000 +0100   Occupancy         max    0.0   
1        2025-04-01 13:59:59.000 +0100   Occupancy         max    0.0   
2        2025-04-01 13:59:59.000 +01

In [ ]:
time_metrics.head()

,id,frequency,start_time,end_time,metric_name,aggregation,value,geometry_id,is_holiday,is_valid,is_working,hierarchy_id,portfolio_id,data_source_id
0,20043506,hourly,2025-04-01 13:00:00.000 +0100,2025-04-01 13:59:59.000 +0100,Occupancy,max,0.0,23209,False,True,True,NaN,23457,1
1,20043507,hourly,2025-04-01 13:00:00.000 +0100,2025-04-01 13:59:59.000 +0100,Occupancy,max,0.0,23210,False,True,True,NaN,23457,1
2,20043508,hourly,2025-04-01 13:00:00.000 +0100,2025-04-01 13:59:59.000 +0100,Occupancy,max,0.0,23211,False,True,True,NaN,23457,1
3,20043509,hourly,2025-04-01 13:00:00.000 +0100,2025-04-01 13:59:59.000 +0100,Occupancy,max,0.0,23212,False,True,True,NaN,23457,1
4,20043510,hourly,2025-04-01 13:00:00.000 +0100,2025-04-01 13:59:59.000 +0100,Occupancy,max,0.0,23213,False,True,True,NaN,23457,1


In [ ]:
time_h = pd.read_csv("/content/metrics_app_hierarchy.csv")
print(time_h)


                   name         display_name hierarchy_type  geometry_id  \
0              building             Building          space        23175   
1        seminar-room-9       Seminar Room 9          space        23611   
2        seminar-room-2       Seminar Room 2          space        23605   
3        seminar-room-3       Seminar Room 3          space        23606   
4              the-hive             The Hive          space        23607   
5        seminar-room-6       Seminar Room 6          space        23608   
6       teaching-room-5      Teaching Room 5          space        23609   
7    specialist-room-15   Specialist Room 15          space        23616   
8       teaching-room-4      Teaching Room 4          space        23610   
9    specialist-room-14   Specialist Room 14          space        23617   
10     teaching-room-10     Teaching Room 10          space        23612   
11     teaching-room-11     Teaching Room 11          space        23613   
12   special

In [ ]:
time_metrics.shape #Counting no of rows and columns in each dataset

(2411456, 14)

In [ ]:
time_h.shape

(32, 10)

In [ ]:
uni_con = time_metrics.nunique(dropna=False).sort_values(ascending=False)
uni_con

,0
id,2411456
value,42353
end_time,3072
start_time,3072
geometry_id,152
metric_name,30
data_source_id,4
aggregation,4
frequency,2
is_working,2


In [ ]:
uni_con_hier = time_h.nunique(dropna=False).sort_values(ascending=False)
uni_con_hier

,0
name,32
display_name,32
geometry_id,32
id,32
hierarchy_type,1
portfolio_name,1
parent_id,1
longitude,1
latitude,1
portfolio_id,1


By Using Unique fuction we understand that:

In hierarchy dataset contains 32 unique room/space records.Each record has a unique in 4 column and remaining 6 have same or nul values which can be removed without data loss.

In Time metrics data  contains 25 unique metric type and 150 unique geometry identifiers and 4 columns has same or nul values.

In [ ]:
print("Time Metrics data")
print("-----------------")
print(time_metrics.isnull().sum()) #checking the null values in both datasets

print("\nHierarchy data")
print("--------------")
print(time_h.isnull().sum())

Time Metrics data
-----------------
id                      0
frequency               0
start_time              0
end_time                0
metric_name             0
aggregation             0
value                   0
geometry_id             0
is_holiday              0
is_valid                0
is_working              0
hierarchy_id      2411456
portfolio_id            0
data_source_id          0
dtype: int64

Hierarchy data
--------------
name               0
display_name       0
hierarchy_type     0
geometry_id        0
portfolio_name     0
id                 0
parent_id          0
longitude         32
latitude          32
portfolio_id       0
dtype: int64


In [ ]:
time_metrics.drop(columns=["hierarchy_id","is_holiday","is_valid"], inplace = True) #removing null value- columns from the dataset

time_h.drop(columns=["longitude","latitude"], inplace = True)

In [ ]:
time_metrics["start_time"] = pd.to_datetime(time_metrics["start_time"],errors="coerce",utc= True)
time_metrics["end_time"] = pd.to_datetime(time_metrics["end_time"],errors="coerce",utc= True)

In [ ]:
time_metrics["start_month"] = time_metrics["start_time"].dt.month
time_metrics["start_month_name"] = time_metrics["start_time"].dt.month_name()
time_metrics["start_hour"] = time_metrics["start_time"].dt.hour



In [ ]:
time_metrics["month"] = time_metrics["start_time"].dt.month_name()
time_metrics["hour"] = time_metrics["start_time"].dt.hour

In [ ]:
time_metrics[["start_time","end_time","month","hour"]].head()

,start_time,end_time,month,hour
0,2025-04-01 12:00:00+00:00,2025-04-01 12:59:59+00:00,April,12
1,2025-04-01 12:00:00+00:00,2025-04-01 12:59:59+00:00,April,12
2,2025-04-01 12:00:00+00:00,2025-04-01 12:59:59+00:00,April,12
3,2025-04-01 12:00:00+00:00,2025-04-01 12:59:59+00:00,April,12
4,2025-04-01 12:00:00+00:00,2025-04-01 12:59:59+00:00,April,12


In [ ]:
time_metrics["time_difference"] = time_metrics["end_time"] - time_metrics["start_time"]

time_metrics["time_difference_hr"] = ((time_metrics["end_time"] - time_metrics["start_time"]).dt.total_seconds()/3600).round(2)

In [ ]:
time_metrics[["start_time", "end_time", "time_difference","time_difference_hr"]].head()

,start_time,end_time,time_difference,time_difference_hr
0,2025-04-01 12:00:00+00:00,2025-04-01 12:59:59+00:00,0 days 00:59:59,1.0
1,2025-04-01 12:00:00+00:00,2025-04-01 12:59:59+00:00,0 days 00:59:59,1.0
2,2025-04-01 12:00:00+00:00,2025-04-01 12:59:59+00:00,0 days 00:59:59,1.0
3,2025-04-01 12:00:00+00:00,2025-04-01 12:59:59+00:00,0 days 00:59:59,1.0
4,2025-04-01 12:00:00+00:00,2025-04-01 12:59:59+00:00,0 days 00:59:59,1.0


In [ ]:
time_metrics["time_difference_hr"].unique()

array([ 1., 24.])

In [ ]:
essential_cols = ["end_time", "metric_name","aggregation","value","geometry_id","is_working","portfolio_id","data_source_id"]

time_metrics = time_metrics.dropna(subset= essential_cols)
time_metrics = time_metrics.reset_index(drop=True)

In [ ]:
time_metrics.head()

,id,frequency,start_time,end_time,metric_name,aggregation,value,geometry_id,is_working,portfolio_id,data_source_id,start_month,start_month_name,start_hour,month,hour,time_difference,time_difference_hr
0,20043506,hourly,2025-04-01 12:00:00+00:00,2025-04-01 12:59:59+00:00,Occupancy,max,0.0,23209,True,23457,1,4,April,12,April,12,0 days 00:59:59,1.0
1,20043507,hourly,2025-04-01 12:00:00+00:00,2025-04-01 12:59:59+00:00,Occupancy,max,0.0,23210,True,23457,1,4,April,12,April,12,0 days 00:59:59,1.0
2,20043508,hourly,2025-04-01 12:00:00+00:00,2025-04-01 12:59:59+00:00,Occupancy,max,0.0,23211,True,23457,1,4,April,12,April,12,0 days 00:59:59,1.0
3,20043509,hourly,2025-04-01 12:00:00+00:00,2025-04-01 12:59:59+00:00,Occupancy,max,0.0,23212,True,23457,1,4,April,12,April,12,0 days 00:59:59,1.0
4,20043510,hourly,2025-04-01 12:00:00+00:00,2025-04-01 12:59:59+00:00,Occupancy,max,0.0,23213,True,23457,1,4,April,12,April,12,0 days 00:59:59,1.0


In [ ]:
time_metrics_clean = time_metrics.copy()
hierarchy_clean = time_h.copy()

In [ ]:
merged_data = time_metrics_clean.merge(
    hierarchy_clean[["geometry_id","display_name","name","hierarchy_type","portfolio_name"]],
    on = "geometry_id",
    how= "left",
    indicator=True
)

In [ ]:
print(merged_data["_merge"].value_counts())

join_success = (merged_data["_merge"] == "both").mean() * 100
print("join success percentage:", round(join_success,2),"%")

_merge
both          1391947
left_only     1019509
right_only          0
Name: count, dtype: int64
join success percentage: 57.72 %


Above output tells:

- Around 57.72% of metric records successfully matched with hierarchy dataset using geometry_id.

- remaining 42.26% of metric records did not match any room name. this means

- some geometry_id values in the metrics are not room-level spaces they inclued:

   - Building- level recodes
   - extrnal/ weather metrics
   - sensor - level recods
So chatbot needs validation step before giving room-level answers.    

In [ ]:
merged_data["mapping_status"] = merged_data["_merge"].map({
    "both": "mapped to room",
    "left_only" : "Unmapped",
    "right_only" : "Hierarchy only"
})

In [ ]:
merged_data["mapping_status"].value_counts()

,count
mapping_status,
mapped to room,1391947
Unmapped,1019509
Hierarchy only,0


In [ ]:
metric_mapping_summary = (
    merged_data
    .groupby(["metric_name", "mapping_status"])
    .size()
    .unstack(fill_value=0)
    .reset_index()
)

metric_mapping_summary["total"] = metric_mapping_summary.sum(axis=1,numeric_only=True)

metric_mapping_summary["mapped_percent"] = (
    metric_mapping_summary.get("mapped to room",0)/metric_mapping_summary["total"]*100
).round(2)

display(metric_mapping_summary.sort_values("mapped_percent", ascending= False))

/tmp/ipykernel_816/1341334277.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  .groupby(["metric_name", "mapping_status"])


mapping_status,metric_name,Unmapped,Hierarchy only,mapped to room,total,mapped_percent
1,OccupancyOptimised,0,0,81,81,100.00
2,ambientNoise,0,0,3009,3009,100.00
4,cloudcover,0,0,1446,1446,100.00
8,extHumidity,0,0,1446,1446,100.00
7,equilibriumMoistureContent,0,0,105903,105903,100.00
6,daysToMold,0,0,105903,105903,100.00
10,feelslike,0,0,1446,1446,100.00
9,extTemp,0,0,1446,1446,100.00
24,soundLevel,0,0,3009,3009,100.00
26,tvoc,0,0,3009,3009,100.00


Good mapping for our model:

temp - 98.20%

humidity - 98.20%

CO2 - 98.16%

Occupancy - 88.26%

PeopleMotion - 54.60%



PeopleCount - 0% , so we can't answer which room has higest peopleCount?

Creating Room-level Dataset

In [ ]:
room_level_data = merged_data[
    merged_data["mapping_status"] == "mapped to room"
].copy()

print(room_level_data.shape)

(1391947, 24)


In [ ]:
room_level_data["metric_name"].value_counts()

,count
metric_name,
peopleMotion,210676
Occupancy,174741
humidity,117768
temp,117744
co2,114843
daysToMold,105903
equilibriumMoistureContent,105903
batteryLevel,105903
metalCorrosion,105903


In [ ]:
all_metrics_clean = merged_data.copy()

room_level_metrics = merged_data[
    merged_data["display_name"].notna()
].copy()

unmapped_metrics = merged_data[
    merged_data["display_name"].isna()
].copy()

In [ ]:
print("Original cleaned rows:", len(time_metrics_clean))
print("Merged rows:", len(merged_data))

assert len(time_metrics_clean) == len(merged_data), \
"The hierarchy merge has change the number of records."

Original cleaned rows: 2411456
Merged rows: 2411456


In [ ]:
print("Hierarchy geometry_id is unique:",
      hierarchy_clean["geometry_id"].is_unique
      )

Hierarchy geometry_id is unique: True


In [ ]:
room_level_metrics.to_csv(
    "../room_level_metrics.csv",
    index= False
)

unmapped_metrics.to_csv(
    "../unmapped_metrics.csv",
    index=False
)

metric_mapping_summary.to_csv(
     "../metric_mapping_summary.csv",
     index=False
)

**Creating configuration dictionary**

In [ ]:
SUPPORTED_ROOM_METRICS = {
    "co2": {
        "unit" : "ppm",
        "default_aggregation": "mean",
        "chart_types": ["line", "bar"]
    },
    "temp": {
        "unit": "C",
        "default_aggregation": "mean",
        "chart_types": ["line","bar"]
    },
    "humidity": {
        "unit": "%",
        "default_agregation": "mean",
        "chart_types": ["line","bar"]
    },
    "Occupancy": {
        "unit": "occupancy value",
        "default_agregation": "mean",
        "chart_types": ["line","bar"]
}
}

This becomes the first validation layer for the chartbot.

**Funtion 1: Rank rooms**

In [ ]:
def rank_rooms(
    data,
    metric_name,
    start_date=None,
    end_date=None,
    aggregation="mean",
    top_n = 10,
    assending= False
):
    df = data[
        (data["metric_name"] == metric_name) &
        (data["aggregation"] == aggregation) &
        (data["frequency"] == "houly")
    ].copy()

    if start_date is not None:
      df = df[df["start_time"] >= pd.to_datetime(start_date)]

    if end_date is not None:
      df = df[df["start_name"] <= pd.to_datetime(end_date)]

    result = (
        df.groupby("display_name", as_index= False)["value"]
        .mean()
        .sort_value("value", ascending=ascending)
        .head(top_n)
    )

    return result

**Function 2: Compare rooms**

In [ ]:
def compare_rooms(
    data,
    room_names,
    metric_name,
    aggregation="mean",
    frequency="hourly"
):
    # Standardise the requested room names
    requested_rooms = [
        room.strip().casefold()
        for room in room_names
    ]

    df = data[
        data["display_name"]
        .astype(str)
        .str.strip()
        .str.casefold()
        .isin(requested_rooms)
        &
        (
            data["metric_name"]
            .astype(str)
            .str.strip()
            .str.casefold()
            == metric_name.strip().casefold()
        )
        &
        (
            data["aggregation"]
            .astype(str)
            .str.strip()
            .str.casefold()
            == aggregation.strip().casefold()
        )
        &
        (
            data["frequency"]
            .astype(str)
            .str.strip()
            .str.casefold()
            == frequency.strip().casefold()
        )
    ].copy()

    if df.empty:
        print("No matching records found.")
        print("Check the room names, metric, aggregation and frequency.")
        return pd.DataFrame()

    result = (
        df.groupby("display_name", as_index=False)
        .agg(
            average_value=("value", "mean"),
            minimum_value=("value", "min"),
            maximum_value=("value", "max"),
            record_count=("value", "count")
        )
        .sort_values("average_value", ascending=False)
    )

    return result

In [ ]:
room_comparison = compare_rooms(
    room_level_data,
    room_names=[
        "Seminar Room 2",
        "Seminar Room 3"


    ],
    metric_name="co2",
    aggregation="mean",
    frequency="hourly"
)

display(room_comparison)


,display_name,average_value,minimum_value,maximum_value,record_count
0,Seminar Room 2,590.855979,380.000000,2269.0,2210
1,Seminar Room 3,552.514385,380.666667,4242.0,2254


In [ ]:
co2_rooms = (
    room_level_data.loc[
        room_level_data["metric_name"].str.lower() == "co2",
        ["display_name", "name"]
    ]
    .drop_duplicates()
    .sort_values("display_name")
)

display(co2_rooms)

,display_name,name
9169,Cafe,cafe
14463,Dining Room,dining-room
14461,Founders Room,founders-room
9167,Lecture Theatre 4,lecture-theatre-4
9168,Lecture Theatre 5,lecture-theatre-5
9174,Library 1st Floor,library-1st-floor
9175,Library 2nd Floor,library-2nd-floor
9165,Meeting Room A,meeting-room-a
10162,Meeting Room B,meeting-room-b
9166,Office Room 11A,office-room-11a


In [ ]:
import plotly.express as px

fig = px.bar(
    room_comparison,
    x="display_name",
    y="average_value",
    title="Average CO₂ Comparison by Room",
    labels={
        "display_name": "Room",
        "average_value": "Average CO₂ Value"
    },
    text_auto=".1f"
)

fig.show()

**Function 3: Metric trend for one room**

In [ ]:
def room_metric_trend(
    data,
    room_name,
    metric_name,
    start_date = None,
    end_date=None,
    aggregation="mean",
    top_n = 10,
    assending= "mean",
    frequency="hourly"
):

    df = data[
         (data["display_name"].str.lower() == room_name.lower()) &
         (data["metric_name"] == metric_name)&
         (data["aggregation"] == aggregation) &
         (data["frequency"] == frequency)
    ].copy()

    if start_date is not None:
      df = df[df["start_time"] <= pd.to_datetime(start_date, utc=True)]

    if end_date is not None:
       df = df[df["start_time"] <= pd.to_datetime(end_date, utc=True)]

    return df.sort_values("start_time")[
        ["start_time", "display_name","metric_name","aggregation","frequency", "value"]
    ]



In [ ]:
room_level_data.loc[
    room_level_data["display_name"]
    .astype(str)
    .str.contains("seminar", case=False, na=False),
    ["display_name","name"]
].drop_duplicates()

,display_name,name
9097,Seminar Room 2,seminar-room-2
9098,Seminar Room 3,seminar-room-3
9500,Seminar Room 6,seminar-room-6
9510,Seminar Room A,seminar-room-a
11249,Seminar Room 9,seminar-room-9


In [ ]:
trend_result = room_metric_trend(
    room_level_data,
    room_name= "Seminar Room 3",
    metric_name="co2",
    aggregation="mean",
    frequency= "hourly"
)

display(trend_result.head(20))
print("number of records:", len(trend_result))

,start_time,display_name,metric_name,aggregation,frequency,value
66814,2025-04-01 00:00:00+00:00,Seminar Room 3,co2,mean,hourly,903.166667
66584,2025-04-01 01:00:00+00:00,Seminar Room 3,co2,mean,hourly,856.166667
66263,2025-04-01 02:00:00+00:00,Seminar Room 3,co2,mean,hourly,822.333333
66211,2025-04-01 03:00:00+00:00,Seminar Room 3,co2,mean,hourly,801.666667
65965,2025-04-01 04:00:00+00:00,Seminar Room 3,co2,mean,hourly,774.666667
65796,2025-04-01 05:00:00+00:00,Seminar Room 3,co2,mean,hourly,769.666667
65594,2025-04-01 06:00:00+00:00,Seminar Room 3,co2,mean,hourly,758.166667
65435,2025-04-01 07:00:00+00:00,Seminar Room 3,co2,mean,hourly,764.333333
65244,2025-04-01 08:00:00+00:00,Seminar Room 3,co2,mean,hourly,783.833333
65135,2025-04-01 09:00:00+00:00,Seminar Room 3,co2,mean,hourly,1055.250000


number of records: 2254


In [ ]:
import plotly.express as px

fig = px.line(
    trend_result,
    x= "start_time",
    y="value",
    title = "Hourly Mean CO2 trend - Seminar Room 3",
    labels={
        "start_time": "Date and Time",
        "value": "CO2 Level"
    }

)

fig.update_layout(
    xaxis_title = "Date and Time",
    yaxis_title= "Mean CO2 Level",
    hovermode= "x unified"
)

fig.show()

In [ ]:
trend_check = trend_result.sort_values("start_time").copy()

print("Number of records:", len(trend_check))
print("Duplicte timestamps:", trend_check["start_time"].duplicated().sum())
print("Missing values", trend_check["value"].isna().sum())

print("\nNumerical summary:")
display(trend_check["value"].describe())

print("\nLargest 10 readings:")
display(
    trend_check.nlargest(10, "value")[
        ["start_time","display_name","value"]
    ]
)

Number of records: 2254
Duplicte timestamps: 0
Missing values 0

Numerical summary:


,value
count,2254.000000
mean,552.514385
std,304.117433
min,380.666667
25%,382.000000
50%,399.500000
75%,613.250000
max,4242.000000



Largest 10 readings:


,start_time,display_name,value
63983,2025-04-01 15:00:00+00:00,Seminar Room 3,4242.000000
64234,2025-04-01 14:00:00+00:00,Seminar Room 3,3740.750000
64270,2025-04-01 13:00:00+00:00,Seminar Room 3,3093.454545
64474,2025-04-01 12:00:00+00:00,Seminar Room 3,2703.666667
60486,2025-04-02 10:00:00+00:00,Seminar Room 3,2543.166667
47231,2025-04-05 12:00:00+00:00,Seminar Room 3,2497.750000
15574,2025-04-08 17:00:00+00:00,Seminar Room 3,2360.000000
63754,2025-04-01 16:00:00+00:00,Seminar Room 3,2314.583333
64763,2025-04-01 11:00:00+00:00,Seminar Room 3,2152.636364
15400,2025-04-08 18:00:00+00:00,Seminar Room 3,2129.500000


In [ ]:
fig = px.line(
    daily_co2_summary,
    x="start_time",
    y=["daily_mean", "daily_max"],
    title="Daily Mean and Maximum CO₂ Trend — Seminar Room 3",
    labels={
        "start_time": "Date",
        "value": "CO₂ Value",
        "variable": "Measure"
    }
)

fig.for_each_trace(
    lambda trace: trace.update(
        name={
            "daily_mean": "Daily Mean",
            "daily_max": "Daily Maximum"
        }.get(trace.name, trace.name)
    )
)

fig.update_layout(
    hovermode="x unified",
    legend_title_text="Measure"
)

fig.show()

This Room test proved:

metric validated

matching records filtered

hourly data returned

dialy aggregation calculated

graph generated

result interpreted



In [ ]:
def available_metrics_for_room(data, room_name):
    result = (
        data[
            data["display_name"]
            .astype(str)
            .str.strip()
            .str.casefold()
            == room_name.strip().casefold()
        ]
        .groupby(["metric_name", "aggregation", "frequency"])
        .size()
        .reset_index(name="record_count")
        .sort_values(["metric_name", "aggregation"])
    )

    return result

In [ ]:
display(
    available_metrics_for_room(
        room_level_data,
        "seminar Room 9"
    )
)

,metric_name,aggregation,frequency,record_count
0,Occupancy,max,hourly,1814
1,Occupancy,mean,hourly,1814
2,Occupancy,min,hourly,1814
3,OccupancyOptimised,max,hourly,1
4,OccupancyOptimised,mean,hourly,1
5,OccupancyOptimised,min,hourly,1
6,ambientNoise,max,hourly,59
7,ambientNoise,mean,hourly,59
8,ambientNoise,min,hourly,59
9,humidity,max,hourly,59


This tells chatbot which metrics can actually be analysed for that room.

**Function 4: Threshold analysis**

In [ ]:
def rooms_above_threshold(
    data,
    metric_name,
    threshold,
    aggregation="mean"
):
    df = data[
        (data["metric_name"] == metric_name) &
        (data["aggregation"] == aggregation) &
        (data["frequency"] == "hourly")
    ].copy()

    result = (
        df.groupby("display_name", as_index=False)["value"]
        .mean()
    )

    return result[
        result["value"] > threshold
    ].sort_values("value", ascending=False)

This functions are real analytical engine that LLM should only decide which fuction and parameter to use

Defining structured query format

In [ ]:
query_request = {
    "intent": "rank_rooms",
    "metric": "co2",
    "rooms": None,
    "start_date": "2025-06-01",
    "end_date": "2025-06-30",
    "aggregation": "mean",
    "frequency": "hourly",
    "chart_type": "bar",
    "top_n": 10
}